hilti/include/rt/type-info.h

```diff
@@ -0,0 +1,1071 @@
+        std::vector<std::pair<const struct_::Field&, Value>> values;
+
+        for ( const auto& f : _fields ) {
+            auto x = Value(static_cast<const char*>(v.pointer()) + f.offset, f.type, v);
+            values.emplace_back(f, f.value(x));
+        }
+
+        return values;
+    }
+
+private:
+    std::vector<struct_::Field> _fields;
+};
+
+/** Auxiliary type information for type ``time`. */
+class Time : public detail::AtomicType<hilti::rt::Time> {};
+
+class Tuple;
+
+namespace tuple {
+
+/** Auxiliary type information for type ``tuple`` describing one tuple element. */
+class Element {
+public:
+    /**
+     * Constructor.
+     *
+     * @param name ID of the element, with an empty string indicating no name
+     * @param type type of the field
+     * @param offset offset of the field inside the tuple
+     */
+    Element(const char* name, const TypeInfo* type, std::ptrdiff_t offset) : name(name), type(type), offset(offset) {}
+
+    const std::string name; /**< ID of the element, with an empty string indicating no name */
+    const TypeInfo* type;   /**< type of the element */
+
+private:
+    friend class type_info::Tuple;
+
+    const std::ptrdiff_t offset;
+};
+
+}; // namespace tuple
+
+/** Auxiliary type information for type ``tuple`. */
+class Tuple {
+public:
+    /**
+     * Constructor
+     *
+     * @param labels the tuple's elements
+     */
+    Tuple(std::vector<tuple::Element> elements) : _elements(std::move(elements)) {}
+
+    /** Returns the tuple's elements. */
+    const auto& elements() const { return _elements; }
+
+    /**
+     * Returns a vector that can be iterated over to visit all the elements.
+     *
+     * @param v the value referring to the tuple to iterate over
+     *
+     * @return a vector of pairs ``(element, value)`` where *element* is the
+     * current ``tuple::Element` and *value* is the element's value.
+     */
+    auto iterate(const Value& v) const {
+        std::vector<std::pair<const tuple::Element&, Value>> values;
+
+        for ( const auto& f : _elements )
+            values.emplace_back(f, Value(static_cast<const char*>(v.pointer()) + f.offset, f.type, v));
+
+        return values;
+    }
+
+private:
+    std::vector<tuple::Element> _elements;
+};
+
+namespace union_ {
+
+/** Auxiliary type information for type ``union`` describing one field. */
+struct Field {
+    /**
+     * Constructor.
+     *
+     * @param name ID of the field
+     * @param type type of the field
+     */
+    Field(const char* name, const TypeInfo* type) : name(name), type(type) {}
+
+    const std::string name; /**< ID of the field */
+    const TypeInfo* type;   /**< type of the field */
+};
+
+}; // namespace union_
+
+/** Auxiliary type information for type ``union`. */
+class Union {
+public:
+    /**
+     * Type of a function that, given a union value, returns the index of the
+     * currently set field, with -1 indicating no field being set.
```

Comment:
Since this function returns a `size_t` it should be impossible to return `-1` from it. Can we change this to return e.g., an `optional<size_t>` instead?

Reply:
It's passing through std::variant_npos, I'll add a local const mapping to that.